src/main/java/com/google/sps/servlets/EnterQueue.java

```diff
@@ -46,19 +49,70 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
 
     try {
       String classCode = request.getParameter("classCode").trim();
-      Key classKey = KeyFactory.stringToKey(classCode);
-      Entity classEntity = datastore.get(classKey);
 
-      String idToken = request.getParameter("idToken");
-      FirebaseToken decodedToken = authInstance.verifyIdToken(idToken);
+      if (request.getParameter("enterTA") == null) {
+        Key classKey = KeyFactory.stringToKey(classCode);
+        Entity classEntity = datastore.get(classKey);
 
-      ArrayList<String> updatedQueue = (ArrayList) classEntity.getProperty("studentQueue");
-      updatedQueue.add(decodedToken.getUid());
-      classEntity.setProperty("studentQueue", updatedQueue);
+        String idToken = request.getParameter("idToken");
+        FirebaseToken decodedToken = authInstance.verifyIdToken(idToken);
 
-      datastore.put(classEntity);
+        ArrayList<String> updatedQueue = (ArrayList) classEntity.getProperty("studentQueue");
+        String uID = decodedToken.getUid();
+        if (!updatedQueue.contains(uID)) {
+          int retries = 10;
+          while (true) {
+            TransactionOptions options = TransactionOptions.Builder.withXG(true);
+            Transaction txn = datastore.beginTransaction(options);
```

Comment:
I think you'll want to start the transaction earlier (before line 55). You want to make sure that getting the entity is part of the atomic operation.

Reply:
Also make sure to add the transaction to the get call: `datastore.get(txn, classKey);`

Reply:
I looked at the documentation for transactions and it seems like you don't need to explicitly add the txn to the get call